Fixes https://github.com/NuGet/Home/issues/2420

jainaashish opened the pull request:
Updated FindCircularDependency to maintain cache along with BFS traversal

Commit:
Fixes https://github.com/NuGet/Home/issues/2420
Updated FindCircularDependency to maintain cache along with BFS traversal

dnfclas commented:
Hi **@jainaashish**, I'm your friendly neighborhood .NET Foundation Pull Request Bot (You can call me DNFBOT). Thanks for your contribution!
    <span>
        In order for us to evaluate and accept your PR, we ask that you sign a contribution license agreement. It's all electronic and will take just minutes. I promise there's no faxing. https://cla2.dotnetfoundation.org.
    </span>

TTYL, DNFBOT;

emgarten commented on src/NuGet.Core/NuGet.Resolver/PackageResolver.cs:
```diff
@@ -131,7 +131,12 @@ public IEnumerable<PackageIdentity> Resolve(PackageResolverContext context, Canc
                 if (nonAbsentCandidates.Any())
                 {
-                    var circularReferences = ResolverUtility.FindCircularDependency(solution);
+                    //topologicallly sort non absent packages
```
topologica**ll**y

emgarten commented on src/NuGet.Core/NuGet.Resolver/PackageResolver.cs:
```diff
@@ -131,7 +131,12 @@ public IEnumerable<PackageIdentity> Resolve(PackageResolverContext context, Canc
+                    //topologicallly sort non absent packages
+                    var sortedSolution = ResolverUtility.TopologicalSort(nonAbsentCandidates);
+
+                    //Find circular dependency for topologically sorted non absent packages since it will help maintain cache of 
```
nit: spacing on comment

emgarten commented on src/NuGet.Core/NuGet.Resolver/PackageResolver.cs:
```diff
@@ -131,7 +131,12 @@ public IEnumerable<PackageIdentity> Resolve(PackageResolverContext context, Canc
                 {
-                    var circularReferences = ResolverUtility.FindCircularDependency(solution);
+                    //topologicallly sort non absent packages
+                    var sortedSolution = ResolverUtility.TopologicalSort(nonAbsentCandidates);
```
I would expect this sort to be done inside ResolverUtility.FindCircularDependency, I don't see it being used anywhere outside of that method.

emgarten commented on src/NuGet.Core/NuGet.Resolver/ResolverUtility.cs:
```diff
@@ -334,50 +334,125 @@ internal static bool ShouldRejectPackagePair(ResolverPackage p1, ResolverPackage
         public static IEnumerable<ResolverPackage> FindCircularDependency(IEnumerable<ResolverPackage> solution)
         {
+            // to keep track of visited packages to avoid processing them again
+            var visitedPackages = new Dictionary<string, HashSet<ResolverPackage>>();
```
Add a comment that this should be sorted topologically first, it's not clear within that method that this has been optimized before passing it in

emgarten commented on src/NuGet.Core/NuGet.Resolver/ResolverUtility.cs:
```diff
@@ -334,50 +334,125 @@ internal static bool ShouldRejectPackagePair(ResolverPackage p1, ResolverPackage
         }
 
-        private static IEnumerable<ResolverPackage> FindCircularDependency(ResolverPackage package, IEnumerable<string> parents, IEnumerable<ResolverPackage> solution)
+        private static IEnumerable<ResolverPackage> FindCircularDependency(ResolverPackage package, IEnumerable<ResolverPackage> solution, Dictionary<string, HashSet<ResolverPackage>> visitedPackages)
```
Instead of taking `IEnumerable<ResolverPackage>` and looping each time to find the dependency, index this once outside of this method into a: `SortedDictionary<string, SortedSet<string>>` to get package id -> dependency ids   this will also take care of the sorting for you so that this is deterministic. Use StringComparer.Ordinal for all comparers.

emgarten commented on src/NuGet.Core/NuGet.Resolver/ResolverUtility.cs:
```diff
@@ -334,50 +334,125 @@ internal static bool ShouldRejectPackagePair(ResolverPackage p1, ResolverPackage
+                        var dependencyPackage = solution.FirstOrDefault(solutionPackage => StringComparer.OrdinalIgnoreCase.Equals(solutionPackage.Id, dependency.Id));
+                        //If already visited, then check current source package for circular dependency since it means we've already processed child nodes
+                        //bottom-up approach
+                        HashSet<ResolverPackage> value = null;                        
```
nit: Use a descriptive name instead of value

emgarten commented:
:watch: I think this is doing work than it needs to by tracking all of the closures, this method only needs to track circular dependencies and then find the parents of those circles. 

This should be able to run in < 1000ms

Commit:
Fixes https://github.com/NuGet/Home/issues/2420
Updated FindCircularDependency to maintain cache along with BFS traversal.
Further improved performance by optimizing cache.

jainaashish commented:
Removed tracking all the closures, and now only tracking nodes without any circular dependency for future reference.

Also improved performance and now it takes <100ms to complete.

emgarten commented:
< 100ms! :tada: :balloon:

emgarten commented on src/NuGet.Core/NuGet.Resolver/ResolverUtility.cs:
```diff
@@ -330,54 +330,98 @@ internal static bool ShouldRejectPackagePair(ResolverPackage p1, ResolverPackage
-            {
-                // walk the dependencies
-                foreach (var dependency in package.Dependencies.OrderBy(d => d.Id, StringComparer.OrdinalIgnoreCase))
+            //ashish: removed 20 packages check to go by any depth since no more stack processing
```
These comments on the 20 limit can go away

emgarten commented:
:shipit:

emgarten commented:
//cc @zhili1208 @joelverhagen @alpaix

yishaigalatzer commented on src/NuGet.Core/NuGet.Resolver/ResolverUtility.cs:
```diff
@@ -303,7 +303,7 @@ public static IEnumerable<ResolverPackage> TopologicalSort(IEnumerable<ResolverP
             // add any unsorted nodes onto the end of the result
             var sorted = new HashSet<string>(result.Select(n => n.Id), StringComparer.OrdinalIgnoreCase);
-            result.AddRange(nodes.Where(n => !sorted.Contains(n.Id)));
+            result.AddRange(nodes.Where(n => !sorted.Contains(n.Id)).OrderBy(p => p.Id, StringComparer.OrdinalIgnoreCase));
```
nit: rename sorted - because it is not really sorted it just a hashset, which makes the orderby call confusing.

Also move the nodes.... call outside the addrange to a variable - so it is easy to inspect and debug (and you can call that sorted, because now it is)

yishaigalatzer commented on src/NuGet.Core/NuGet.Resolver/ResolverUtility.cs:
```diff
@@ -330,54 +330,98 @@ internal static bool ShouldRejectPackagePair(ResolverPackage p1, ResolverPackage
+            foreach (var package in solution)
+            {                
+                var result = FindCircularDependency(package, packageLookUp, visitedPackages);
+                if (result?.Any() == true)
```
this only returns the first circular reference (which I think is intentional), but that means the comment and method name should reflect it (so perhaps `FindFirstCircularDependency`), and the comment should say check to find the first package that has a circular dependency.

alpaix replied:
This looks unreadable.

yishaigalatzer commented on src/NuGet.Core/NuGet.Resolver/ResolverUtility.cs:
```diff
@@ -330,54 +330,98 @@ internal static bool ShouldRejectPackagePair(ResolverPackage p1, ResolverPackage
+            if (package != null && !package.Absent && package.Dependencies.Any())
+            {                
+                var queue = new Queue<QueueNode>();
+                //added the first initial node to queue
```
newline before the comment

yishaigalatzer commented on src/NuGet.Core/NuGet.Resolver/ResolverUtility.cs:
```diff
@@ -330,54 +330,98 @@ internal static bool ShouldRejectPackagePair(ResolverPackage p1, ResolverPackage
+            {                
+                var queue = new Queue<QueueNode>();
+                //added the first initial node to queue
+                var qNode = new QueueNode(package, new List<ResolverPackage>());
```
just node

yishaigalatzer commented on src/NuGet.Core/NuGet.Resolver/ResolverUtility.cs:
```diff
@@ -330,54 +330,98 @@ internal static bool ShouldRejectPackagePair(ResolverPackage p1, ResolverPackage
                 {
-                    var dependencyPackage = solution.FirstOrDefault(solutionPackage => StringComparer.OrdinalIgnoreCase.Equals(solutionPackage.Id, dependency.Id));
+                    var source = queue.Dequeue();
+                    //access parent list and add current package as well
```
nit: space after // (all over)

yishaigalatzer commented on src/NuGet.Core/NuGet.Resolver/ResolverUtility.cs:
```diff
@@ -330,54 +330,98 @@ internal static bool ShouldRejectPackagePair(ResolverPackage p1, ResolverPackage
 
-                    if (parents.Contains(dependency.Id, StringComparer.OrdinalIgnoreCase))
+                    // walk the dependencies
+                    foreach (var dependency in source.Package.Dependencies.OrderBy(d => d.Id, StringComparer.OrdinalIgnoreCase))
```
is dependencies an empty enumerable when there are not dependencies?

jainaashish replied:
yes

yishaigalatzer commented on src/NuGet.Core/NuGet.Resolver/ResolverUtility.cs:
```diff
@@ -330,54 +330,98 @@ internal static bool ShouldRejectPackagePair(ResolverPackage p1, ResolverPackage
+                        //bottom-up approach                     
+                        if (!visitedPackages.Contains(dependencyPackage))
+                        {
+                            //top-down approach
```
I don't understand the bottom=up and top-down comments, either clarify or remove

yishaigalatzer commented on src/NuGet.Core/NuGet.Resolver/ResolverUtility.cs:
```diff
@@ -330,54 +330,98 @@ internal static bool ShouldRejectPackagePair(ResolverPackage p1, ResolverPackage
+                            //it means we've processed parent node, n moving down-ward
+                            if (parentPackages.Contains(dependencyPackage))
+                            {
+                                // loop detected
```
circular dependency detected?

yishaigalatzer commented on src/NuGet.Core/NuGet.Resolver/ResolverUtility.cs:
```diff
@@ -330,54 +330,98 @@ internal static bool ShouldRejectPackagePair(ResolverPackage p1, ResolverPackage
+                                parentPackages.Add(dependencyPackage);
+                                return parentPackages;
+                            }
+                            //add child node to Queue to process
```
newline before comment

yishaigalatzer commented on src/NuGet.Core/NuGet.Resolver/ResolverUtility.cs:
```diff
@@ -330,54 +330,98 @@ internal static bool ShouldRejectPackagePair(ResolverPackage p1, ResolverPackage
+
             // end of the walk
-            return Enumerable.Empty<ResolverPackage>();
+            return null;
```
Why not return an empty list, so there is no fear of nullrefs?

yishaigalatzer commented on src/NuGet.Core/NuGet.Resolver/ResolverUtility.cs:
```diff
@@ -330,54 +330,98 @@ internal static bool ShouldRejectPackagePair(ResolverPackage p1, ResolverPackage
+        /// <summary>
+        /// Simple QueueNode class to hold package n it's parent node list together
+        /// </summary>
+        private class QueueNode
```
`PackageAndParent` ?

yishaigalatzer commented on test/NuGet.Core.Tests/NuGet.Resolver.Test/ResolverTests.cs:
```diff
@@ -229,6 +229,46 @@ public void ResolveDependenciesForLargeSet()
+        public void ResolveDependenciesForVeryLargeSet()
+        {
+            // Arrange
+            var target = CreatePackage("Package0", "1.0", new Dictionary<string, string>() {
```
would this test timeout before the fix?

jainaashish replied:
Yeah, this will time out before the fix and now it takes around 18ms

yishaigalatzer commented on test/NuGet.Core.Tests/NuGet.Resolver.Test/ResolverTests.cs:
```diff
@@ -229,6 +229,46 @@ public void ResolveDependenciesForLargeSet()
         }
 
         [Fact]
+        public void ResolveDependenciesForVeryLargeSet()
```
Name the test to explicitly reflect the scenario: VeryLargeSet doesn't say much. Is it:

a. Very deep graph?
b. many packages?
c. both (and then you should split it into two tests)

yishaigalatzer commented:
:watch:

Commit:
Fixes https://github.com/NuGet/Home/issues/2420
Updated FindCircularDependency to maintain cache along with BFS traversal.
Further improved performance by optimizing cache.
Fixed code review comments

emgarten commented:
:shipit:

alpaix commented:
:shipit:

dnfclas commented:
**@jainaashish**, Thanks for signing the contribution license agreement so quickly! Actual humans will now validate the agreement and then evaluate the PR.
<br />Thanks, DNFBOT;
